File: web/personal-finance/src/editability.ts
/**
 * Frontend port of the backend editability state machine. Used for UX hints
 * (lock badges, disabled form controls). The backend is the source of truth —
 * this is only a client-side preview.
 */

import { effectiveNow } from "./simulated_date";

export type EditabilityState = "EDITABLE" | "GRACE" | "LOCKED";

const N_FUTURE_MONTHS = 12;
const GRACE_PERIOD_DAYS = 7;
const BUDGET_TZ = "America/New_York";

interface EtCalendarParts {
    year: number;
    month: number;
    day: number;
}

function nowInEt(now: Date): EtCalendarParts {
    const fmt = new Intl.DateTimeFormat("en-US", {
        timeZone: BUDGET_TZ,
        year: "numeric",
        month: "2-digit",
        day: "2-digit",
    });
    const parts = fmt.formatToParts(now);
    const lookup = (type: string): number => {
        const part = parts.find((p) => p.type === type);
        if (!part) throw new Error(`Missing ET date part: ${type}`);
        return parseInt(part.value, 10);
    };
    return { year: lookup("year"), month: lookup("month"), day: lookup("day") };
}

function toMonthOrdinal(year: number, month: number): number {
    return year * 12 + (month - 1);
}

function parseYearMonth(ym: string): { year: number; month: number } {
    if (!/^\d{4}-\d{2}$/.test(ym)) {
        throw new Error(`Invalid year-month: ${ym}, expected YYYY-MM`);
    }
    return { year: parseInt(ym.slice(0, 4), 10), month: parseInt(ym.slice(5), 10) };
}

export function editability(yearMonth: string, now: Date = effectiveNow()): EditabilityState {
    const et = nowInEt(now);
    const target = parseYearMonth(yearMonth);
    const diff = toMonthOrdinal(target.year, target.month) - toMonthOrdinal(et.year, et.month);

    if (diff === 0 || (diff >= 1 && diff <= N_FUTURE_MONTHS)) {
        return "EDITABLE";
    }
    // Grace: previous month, before midnight ET on day 8 of the current month.
    // Equivalently: today's ET day < (1 + GRACE_PERIOD_DAYS).
    if (diff === -1 && et.day < 1 + GRACE_PERIOD_DAYS) {
        return "GRACE";
    }
    return "LOCKED";
}

File: lambda/src/corderohq/budget/editability.py
"""The state function that decides if the app can change a budget year-month.

Pure functions only. No DynamoDB, no env vars.
"""

from __future__ import annotations

from datetime import datetime
from typing import Literal
from zoneinfo import ZoneInfo

EditabilityState = Literal["EDITABLE", "GRACE", "LOCKED"]

_N_FUTURE_MONTHS = 12
_GRACE_PERIOD_DAYS = 7
_BUDGET_TZ = "America/New_York"

_YEAR_MONTH_LEN = 7  # "YYYY-MM"


def _to_month_ordinal(year: int, month: int) -> int:
    """Convert (year, month) to a single integer for easy month-distance math."""
    return year * 12 + (month - 1)


def _from_month_ordinal(ordinal: int) -> tuple[int, int]:
    return (ordinal // 12, (ordinal % 12) + 1)


def _parse_year_month(year_month: str) -> tuple[int, int]:
    """Parse "YYYY-MM" into (year, month). Raises ValueError on malformed input."""
    if len(year_month) != _YEAR_MONTH_LEN or year_month[4] != "-":
        raise ValueError(f"Invalid year-month: {year_month!r}, expected YYYY-MM")
    return int(year_month[:4]), int(year_month[5:])


def editability(year_month: str, now_utc: datetime) -> EditabilityState:
    """Return the editability state of the given year-month at the given instant.

    `now_utc` must be a timezone-aware datetime in UTC.

    Rules:
    - The current year-month (in BUDGET_TZ) is EDITABLE.
    - Any of the next N_FUTURE_MONTHS year-months after current is EDITABLE.
    - The previous year-month is GRACE if `now_et` is strictly before midnight ET
      on day (1 + GRACE_PERIOD_DAYS) of the current year-month. Otherwise LOCKED.
    - Everything else is LOCKED.

    GRACE is treated identically to EDITABLE for mutation checks; the distinction
    exists so the frontend can surface a "grace expires in X days" banner.
    """
    if now_utc.tzinfo is None:
        raise ValueError("now_utc must be timezone-aware")

    tz = ZoneInfo(_BUDGET_TZ)
    now_et = now_utc.astimezone(tz)

    target_year, target_month = _parse_year_month(year_month)
    target_ord = _to_month_ordinal(target_year, target_month)
    current_ord = _to_month_ordinal(now_et.year, now_et.month)
    diff = target_ord - current_ord

    if diff == 0 or 1 <= diff <= _N_FUTURE_MONTHS:
        return "EDITABLE"

    if diff == -1:
        grace_cutoff_et = datetime(
            now_et.year,
            now_et.month,
            1 + _GRACE_PERIOD_DAYS,
            0,
            0,
            0,
            tzinfo=tz,
        )
        if now_et < grace_cutoff_et:
            return "GRACE"

    return "LOCKED"
